### PR TITLE
QR video scanning

### LINKED ISSUE #364: Login File Scanning with camera
Add camera support to the Keyguard QR Scanner for scanning Login Files via camera.

### PR DESCRIPTION
**Pull request by sisou**
Fixes #364

This PR extends the `QrScanner` library with the ability to scan live video for QR codes, converted from https://github.com/nimiq/qr-scanner.

It then adds a `QrVideoScanner` component to encapsulate the scanner. Also includes already-translated strings, taken from the nimiq/vue-components translations.

~~There is a workaround for the "BarcodeScanner", which is sadly not yet recognized by Typescript and thus cannot be used for types.~~

### CODE REVIEW

File: src/lib/QrScanner.js [hunk 1]
```diff
@@ -1,141 +1,565 @@
...
+            }
+            this.$video.srcObject = null;
+            this._offTimeout = undefined;
+        }, 300);
+    }
+
+    /**
+     * @param {HTMLImageElement | HTMLVideoElement | string} imageOrFileOrUrl
+     * @param {RectArea} [scanRegion]
+     * @param {Promise<Worker>} [givenEngine] // TODO: Or BarcodeDetector
```
Comment (mraveux) [inline src/lib/QrScanner.js]: Is there really no way to use `BarcodeDetector` in JsDoc?

File: src/lib/QrScanner.js [hunk 2]
```diff
@@ -1,141 +1,565 @@
...
+            if (exact) {
+                const exactFacingMode = { exact: facingMode };
+                constraintsToTry.forEach(constraint => {
+                    constraint.facingMode = exactFacingMode;
+                });
+            } else {
+                constraintsToTry.forEach(constraint => {
+                    constraint.facingMode = facingMode;
                 });
             }
```
Comment (mraveux) [inline src/lib/QrScanner.js]: Couldn't this be simplified to the following?
```
constraintsToTry.forEach(constraint => {
    constraint.facingMode = exact ? { exact: facingMode } : facingMode;
})
```

File: src/lib/QrScanner.js [hunk 3]
```diff
@@ -1,141 +1,564 @@
...
+     * @returns {Promise<MediaStream>}
+     */
+    _getMatchingCameraStream(constraintsToTry) {
+        if (!navigator.mediaDevices || constraintsToTry.length === 0) {
+            return Promise.reject('Camera not found.');
+        }
+        return navigator.mediaDevices.getUserMedia({
+            video: constraintsToTry.shift(),
+        }).catch(() => this._getMatchingCameraStream(constraintsToTry));
+    }
```
Comment (mraveux) [inline src/lib/QrScanner.js]: Wouldn't it be cleaner as an async function?
```
async _getMatchingCameraStream(constraintsToTry) {
    if (!navigator.mediaDevices || constraintsToTry.length === 0) {
        return Promise.reject('Camera not found.');
    }
    try {
        return navigator.mediaDevices.getUserMedia({
            video: constraintsToTry.shift(),
        });
    } catch (e) {
        return await this._getMatchingCameraStream(constraintsToTry);
    }
}
```

**Review by mraveux — COMMENTED** [review]
Looks nice, as always 👍 
The only thing that prevent me from approving the PR is that, as discussed, it doesn't seem to work as intended on Chrome android (and maybe other browsers but that'll require more testing) but that should not be a big issue to fix imo.

File: src/lib/QrScanner.js (hunk 1 above)
Comment (sisou) [reply src/lib/QrScanner.js]: It's not included in TS. And probably won't be until there's wider support: https://caniuse.com/?search=barcodedetector

File: src/lib/QrScanner.js (hunk 3 above)
Comment (sisou) [reply src/lib/QrScanner.js]: With a slight change, that the `navigator.mediaDevices.getUserMedia()` function must be `await`ed, but yes, that works.

**Review by mraveux — APPROVED** [review]
Perfect! It now works perfectly on mobile devices as well 😃

**Review by danimoh — APPROVED** [review]
Nice work :+1:
I applied a variety of improvements and a fix for scanning invalid LoginFiles / arbitrary QR codes, see commit history.
Please have a look at the changes.
I also rebased the branch onto master to avoid later merge conflicts as I changed code touching the current LoginFile validation that already supports the encoded wallet label.
As the QrScanner and QrVideoScaner are largely adapted from the `qr-scanner` lib and `@nimiq/vue-components`, I took the freedom, to also add me as a co-author to the commits adding these chunks.
However, I'm also not upset if you object to this and want to force-push it away.

File: src/lib/QrScanner.js (hunk 1 above)
Comment (sisou) [reply src/lib/QrScanner.js]: Daniel added a type definition, so now it's included in the types.

**Pull request merged**